INT: add support for smart pointers, Default implementation, struct init shorthand and local variable to Add struct fields fix

Kobzol opened the pull request:
Hi,
I implemented support for smart pointers, Default impl and shorthand init syntax using local variables with same name and type for the Add struct fields fix (https://github.com/intellij-rust/intellij-rust/issues/2311), .

It would be also nice to auto-import the smart pointers if they are not in scope (if we prefill `Rc::new(0)` and `Rc` is not in scope, the user will have to use autoimport after the fix). Any hint on how could I do that? I guess I could run the autoimport action after the fix, but that seems like a heavyweight solution to me.

- [x] smart pointers
- [x] ? use Default impl
- [x] use local variables for initialization
- [ ] automatically import missing types for filled fields
- [ ] ? make the whole analysis dumber/simpler

Kobzol commented:
Added the possibility to initialize with an existing local variable.

Kobzol commented on src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt:
```diff
@@ -531,7 +531,7 @@ fun processLocalVariables(place: RsElement, processor: (RsPatBinding) -> Unit) {
             val el = v.element
             if (el is RsPatBinding) processor(el)
-            true
+            false
```
I think that this should be false, since with true it only returns the first local variable.

Kobzol commented on src/main/kotlin/org/rust/ide/inspections/fixes/InitializeWithDefaultValueFix.kt:
```diff
@@ -47,6 +49,14 @@ class InitializeWithDefaultValueFix(element: RsElement) : LocalQuickFixAndIntent
         }
     }
 
+    private fun getVisibleBindings(place: RsElement) : HashSet<RsPatBinding> {
```
This is duplicated in AddStructFieldsFix. Where is a good place to put it to avoid the duplication?

vlad20012 commented on src/main/kotlin/org/rust/lang/core/psi/RsDefaultValueBuilder.kt:
```diff
@@ -19,32 +20,63 @@ class RsDefaultValueBuilder(
+
+                var default = this.defaultValue
+                if (item.implLookup.isDefault(ty)) {
+                    default = psiFactory.createAssocFunctionCall("Default", "default", listOf())
```
I don't think it's reasonable to insert `Default::default()` because it's implemented for most of the types but rarely used manually (mostly generic trait magic like `<T: default> T::default()` in some generic function).

Kobzol replied:
It seemed to me like a good default (no pun intended :slightly_smiling_face: ). I agree that it's probably not seen so often in user code, but I'd rather insert `Default::default()` than `()`, which will cause an error. So maybe use the default only as a last resort instead of `()` if possible, but otherwise prefer per-field initialization?

For me this fix would be mostly used in constructors (such as `T::new()`) and there I expect the local variable/parameter initialization will be most useful.

vlad20012 replied:
Maybe it's ok... anyway I'm going to add a separate quickfix to only add missing fields without initializers #3326

Kobzol replied:
That's a good idea, personally I find the Add/Recursively add terminology a bit confusing. It could be changed to Add (just add field names and let the user fill them by moving the caret to each field successively) and Add + fill, which would work like the current recursive approach.

Kobzol replied:
So should this PR wait for #3326?

vlad20012 commented:
> Also changed "".to_string() to "".to_owned() (https://users.rust-lang.org/t/to-string-vs-to-owned-for-string-literals/1441).
Looks like you read the only first message in the thread, that is outdated now. Please, read [this, actual answer](https://users.rust-lang.org/t/to-string-vs-to-owned-for-string-literals/1441/5)

vlad20012 commented on src/main/kotlin/org/rust/lang/core/resolve/KnownItems.kt:
```diff
@@ -109,7 +109,7 @@ class KnownItems(
     val Box: RsStructOrEnumItemElement? get() = findLangItem("owned_box", "alloc")
     val Rc: RsStructOrEnumItemElement? get() = findItem("alloc::rc::Rc")
-    val Arc: RsStructOrEnumItemElement? get() = findItem("alloc::sync::Arc")
+    val Arc: RsStructOrEnumItemElement? get() = findItem("std::sync::Arc")
```
Hmm, does some test failed on this? It can be a problem in a project with `#![no_std]` and `extern crate alloc;`

vlad20012 replied:
Yeah, `findItem("alloc::sync::Arc") ?: findItem("std::arc::Arc")` looks ok

Kobzol commented:
> > Also changed "".to_string() to "".to_owned() (https://users.rust-lang.org/t/to-string-vs-to-owned-for-string-literals/1441).
> > Looks like you read the only first message in the thread, that is outdated now. Please, read [this, actual answer](https://users.rust-lang.org/t/to-string-vs-to-owned-for-string-literals/1441/5)

Ah, I thought that they want to solve it with `specialization`, which didn't land yet AFAIK. But they just overriden the trait method impl, now I get it :) So should I revert it back?

vlad20012 commented:
> So should I revert it back?

Yep :)

Kobzol commented:
Ok, I reverted the `to_owned() change` and used the `Arc` backup for older Rust. The autoimport of `Rc`, `Arc` etc. that I mentioned should be easier after https://github.com/intellij-rust/intellij-rust/pull/3099 lands.

vlad20012 commented on src/main/kotlin/org/rust/ide/inspections/RsFieldInitShorthandInspection.kt:
```diff
@@ -28,12 +28,16 @@ class RsFieldInitShorthandInspection : RsLocalInspectionTool() {
         }
     }
+
+    fun applyFix(field: RsStructLiteralField)
```
Seems to simple to extract. It's ok to copy-paste these 2 lines. Or at least let's move it to static method (with better name) instead of object method

Kobzol replied:
Moved it to a companion object and renamed to `applyShorthandInit`.

Kobzol commented:
> Some tests are failed.
> Also please reduce number of commits by squashing "fixes" into commits where "bug" was made

Could you please point me to the failed tests? I see the CI green at the latest commit (https://github.com/intellij-rust/intellij-rust/commit/fdb0ae36097244e88a8cc03b293bc4c49a0a35c3). I planned to squash the whole PR into a single commit after it's done.

Kobzol commented on src/main/kotlin/org/rust/ide/annotator/fixes/AddStructFieldsFix.kt:
```diff
@@ -51,12 +54,21 @@ class AddStructFieldsFix(
     }
+
+    private fun getVisibleBindings(place: RsElement) : HashSet<RsPatBinding> {
+        val bindings = HashSet<RsPatBinding>()
```
Actually I just need to iterate though this, so I might as well convert it to a `List`. I wanted to ask whether I should extract it to a separate function (it's used both here and in the initialize with default value fix). If it's extracted into some utility space (I don't know where), then it would make more sense to use the `Map` as you say :)

vlad20012 replied:
@ortem I remember something similar in a one of your quickfies. Could you, please, point that place?

vlad20012 commented:
> Could you please point me to the failed tests?
Ok, I started to write that message yesterday, when there really was some failed tests. I'm sorry!

Kobzol commented:
@vlad20012 @mchernyavsky so is this now obsolete because of https://github.com/intellij-rust/intellij-rust/pull/3627? Or should it be kept as a separate fix (fill with manually entered values x fill with default values)?

mchernyavsky commented:
> so is this now obsolete because of #3627?

@Kobzol No, because by default placeholders from #3627 are filled with default values.

> automatically import missing types for filled fields

I think I will do it myself as part of https://github.com/intellij-rust/intellij-rust/pull/3099. Now I am working on https://github.com/intellij-rust/intellij-rust/issues/3586, https://github.com/intellij-rust/intellij-rust/issues/3495 and https://github.com/intellij-rust/intellij-rust/issues/3223.

Kobzol commented:
Ah sorry, I didn't realize that I actually DO index the variable set by name :slightly_smiling_face: I changed it to `Map` and moved the implementation to `RsDefaultValueBuilder` so that it is not duplicated. Squashed all commits into a6f2bb3 and rebased onto the current master.

Kobzol commented:
It conflicted with some changes in master, I rebased it again to resolve it.

mchernyavsky commented:
bors r=vlad20012
